src/apatheia/jobs.nim

```diff
@@ -16,75 +17,164 @@ logScope:
 
 ## This module provides a simple way to submit jobs to taskpools
 ## and getting a result returned via an async future.
-## 
+##
+## Any compatible arguments of `seq[T]` or `string` args passed
+## via the `submit` macro will be converted into the special `OpenArrayHolder[T]` type.
+## The `submit` macro converts these arguments in this object and retains the 
+## memory associated with the original `seq[T]` or `string` object.
+## This greatly simplifies the passing of these these types in `refc`.
+##
+## Note, for `arc` or `orc` GC's this setup will be replaced with a move operation in the future.
+## These GC's also allow greater support for moving GC types across thread boundaries.
+##
+## Currently this module limits support for GC types to ensure `refc` safety.
 ##
 
 type
-  JobId* = uint ## job id, should match `future.id()`
-
-  JobQueue*[T] = ref object
-    ## job queue object
+  JobQueue*[T] = ref object ## job queue object
     queue*: SignalQueue[(JobId, T)]
     futures*: Table[JobId, Future[T]]
     taskpool*: Taskpool
     running*: bool
 
-  JobResult*[T] = object
-    ## hold a job result to be returned by jobs
+  JobResult*[T] = object ## hold the result of a job after it finishes
     id*: JobId
     queue*: SignalQueue[(JobId, T)]
 
+  OpenArrayHolder*[T] = object
+    data*: ptr UncheckedArray[T]
+    size*: int
+
+  SupportedSeqTypes* = byte | SomeInteger | SomeFloat
+
+template toOpenArray*[T](arr: OpenArrayHolder[T]): auto =
+  system.toOpenArray(arr.data, 0, arr.size)
+
+func jobId*[T](fut: Future[T]): JobId =
+  JobId fut.id()
+
 proc processJobs*[T](jobs: JobQueue[T]) {.async.} =
   ## Starts a "detached" async processor for a given job queue.
   ## 
   ## This processor waits for events from the queue in the JobQueue
-  ## and complete the associated futures.
+  ## and completes the associated future.
 
   const tn: string = $(JobQueue[T])
-  info "Processing jobs in job queue for type ", type=tn
+  info "Processing jobs in job queue for type ", type = tn
   while jobs.running:
-    let res = await(jobs.queue.wait()).get()
-    trace "got job result", jobResult = $res
-    let (id, ret) = res
-    var fut: Future[T]
-    if jobs.futures.pop(id, fut):
-      fut.complete(ret)
+    let (id, ret) = await(jobs.queue.wait()).get()
+    trace "got job result", jobId = id
+    releaseMemory(id) # always release any retained memory
+    if (var fut: Future[T]; jobs.futures.pop(id, fut)):
+      if not fut.finished():
+        fut.complete(ret)
```

Comment:
Is the `not fut.finished()` check needed?

Reply:
Good question. If you do a double completion you'll get an exception, but are there acceptable situations in which the user completes the future? I guess in case it gets cancelled?

Reply:
Yah, I'm not sure. I'm going to work on cancellation next. Hmmm, yes a cancelled future that isn't in time to stop the tasklet would end up here. In that case it's been cancelled so just ignore the result basically. 

Thanks!